pkg/util/winutil/scmmonitor.go

```diff
@@ -0,0 +1,203 @@
+	// EnumServiceStatusEx requires only SC_MANAGER_ENUM_SERVICE.  Switch to
+	// new library to use least privilege
+	h, err := windows.OpenSCManager(nil, nil, windows.SC_MANAGER_ENUMERATE_SERVICE)
+	if err != nil {
+		log.Warnf("Failed to connect to scm %v", err)
+		return fmt.Errorf("Failed to open SCM %v", err)
+	}
+	m := &mgr.Mgr{Handle: h}
+	defer m.Disconnect()
+
+	var bytesNeeded, servicesReturned uint32
+	var buf []byte
+	for {
+		var p *byte
+		if len(buf) > 0 {
+			p = &buf[0]
+		}
+		err = windows.EnumServicesStatusEx(m.Handle, windows.SC_ENUM_PROCESS_INFO,
+			windows.SERVICE_WIN32, windows.SERVICE_STATE_ALL,
+			p, uint32(len(buf)), &bytesNeeded, &servicesReturned, nil, nil)
+		if err == nil {
+			break
+		}
+		if err != windows.ERROR_MORE_DATA {
+			return fmt.Errorf("Failed to enum services %v", err)
+		}
+		if bytesNeeded <= uint32(len(buf)) {
+			return err
+		}
+		buf = make([]byte, bytesNeeded)
+	}
+	if servicesReturned == 0 {
+		return nil
+	}
+
+	//var services []windows.ENUM_SERVICE_STATUS_PROCESS
+	//services := (*[1<<29 - 1]windows.ENUM_SERVICE_STATUS_PROCESS)(unsafe.Pointer(&buf[0]))[:servicesReturned]
+	services := unsafe.Slice((*windows.ENUM_SERVICE_STATUS_PROCESS)(unsafe.Pointer(&buf[0])), servicesReturned)
+
+	//hdr.Data = unsafe.Pointer(&buf[0])
+	//hdr.Len = int(servicesReturned)
+	//hdr.Cap = int(servicesReturned)
+
+	newmap := make(map[uint64]*Service)
+	for idx, svc := range services {
+		var thissvc *Service
+		var ok bool
+		if thissvc, ok = newmap[uint64(svc.ServiceStatusProcess.ProcessId)]; !ok {
+			thissvc = &Service{}
+			newmap[uint64(svc.ServiceStatusProcess.ProcessId)] = thissvc
+		}
+		thissvc.essp = append(thissvc.essp, &services[idx])
+	}
+	var current windows.Filetime
+	windows.GetSystemTimeAsFileTime(&current)
+	scm.lastMapTime = uint64(current.Nanoseconds())
+	scm.pidToService = newmap
+	scm.serviceRefreshes++
+	return nil
+}
+
+func (s *Service) toServiceInfo() *ServiceInfo {
+	var si ServiceInfo
+	for _, inf := range s.essp {
+		si.DisplayName = append(si.DisplayName, windows.UTF16PtrToString(inf.DisplayName))
+		si.ServiceName = append(si.ServiceName, windows.UTF16PtrToString(inf.ServiceName))
+	}
+	return &si
+
+}
+
+func (scm *SCMMonitor) getServiceFromCache(pid, pidstart uint64) (*ServiceInfo, error) {
+	var err error
+	if val, ok := scm.pidToService[pid]; ok {
+		// it was in there.  see if it's the right one
+		if val.startTime == 0 {
+			// we don't prepopulate the cache with pid start times
+			val.startTime, err = pGetProcessStartTimeAsNs(pid)
+			if err != nil {
+				return nil, err
+			}
+		}
+		if val.startTime == pidstart {
+			// it's the same one.
+			return val.toServiceInfo(), nil
+		}
+	}
+	return nil, nil
+}
+
+// GetServiceInfo gets the service name and display name if the process identified
+// by the pid is in the SCM.  A process which is not an SCM controlled service will
+// return nil with no error
+func (scm *SCMMonitor) GetServiceInfo(pid uint64) (*ServiceInfo, error) {
```

Comment:
How is this function intended to be used? If used in a `EnumProcess` loop and checking if each PID is a service it seems like it would frequently fall down to the `refreshCache` section, which means it may not be worth having the additional code complexity of the cache.

Reply:
it is exactly intended to be used in an `EnumProcess` loop.  However, the whole point of checking the process start time vs. the last time the table was generated is exactly to prevent that.

If you're in an `EnumProcess` loop, after the first time the cache is regenerated, then the cache will be more recent than the process start times for the remainder of the loop, in which case the refresh would only happen once.  Or, at least that's the intent.

Reply:
Oh right, I was thinking about successive `EnumProcess` loops, I forgot about the multiple calls in the inner loop. 🤦

Reply:
Though it would still be possible to have multiple `refreshCache` calls in a single `EnumProcess` loop, right? Since the refresh is dependent on the start time of the current process being analyzed. Would it be better to separate the creating the cache and getting info from the cache?
So the usage would look like
```
sm = GetServiceMonitor()
sm.GetServiceProcesses()
for process in EnumProcess:
  x = sm.IsService(process.pid)
```